NeurodataWithoutBorders/pynwb
# fix/mpi_io

## Motivation

Add MPI functionality to `NWBHDF5IO`
fix #362

## How to test the behavior?
Creating an automatic test for this will take some work, since it requires HDF5 to be compiled against MPI. For now, you can test with the following code:
```python
from mpi4py import MPI                                                                                                
import h5py                                                                                                           
                                                                                                                      
from dateutil import tz                                                                                               
                                                                                                                      
from pynwb.form.backends.hdf5.h5tools import HDF5IO                                                                   
from pynwb import NWBHDF5IO, NWBFile, TimeSeries                                                                      
from datetime import datetime                                                                                         
                                                                                                                      
start_time = datetime(2018, 4, 25, 2, 30, 3, tzinfo=tz.gettz('US/Pacific'))                                           
                                                                                                                      
fname = 'test_parallel_pynwb.nwb'                                                                                     
                                                                                                                      
rank = MPI.COMM_WORLD.rank  # The process ID (integer 0-3 for 4-process run)                                          
                                                                                                                      
if not rank:                                                                                                          
    nwbfile = NWBFile('aa', 'aa', datetime.now())                                                                     
    nwbfile.add_acquisition(TimeSeries('ts_name', description='desc', data=[0, 0, 0, 0], rate=100., unit='m'))        
    with NWBHDF5IO(fname, 'w') as io:                                                                                 
        io.write(nwbfile)                                                                                             
                                                                                                                      
MPI.COMM_WORLD.Barrier()                                                                                              
                                                                                                                      
with NWBHDF5IO(fname, 'a', comm=MPI.COMM_WORLD) as io:                                                                
        nwbfile = io.read()
        print(rank)                                                                                           
        nwbfile.acquisition['ts_name'].data[rank] = rank
```
Run the above with `mpiexec -n 4 python test_parallel_script.py`
The final nwb file should have `nwbfile.acquisition['ts_name'].data[:] == [0, 1, 2, 3]`

## Checklist
